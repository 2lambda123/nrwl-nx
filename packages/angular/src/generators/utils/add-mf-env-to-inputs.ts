import { type Tree, readNxJson, updateNxJson } from '@nx/devkit';

export function addMfEnvToTargetDefaultInputs(tree: Tree) {
  const nxJson = readNxJson(tree);
  const webpackExecutor = '@nx/angular:webpack-browser';
  const mfEnvVar = 'NX_MF_DEV_SERVER_STATIC_REMOTES';

  let mfEnvVarExists = false;
  if (nxJson.targetDefaults && webpackExecutor in nxJson.targetDefaults) {
    const webpackExecutorTargetDefaults =
      nxJson.targetDefaults[webpackExecutor];

    if (webpackExecutorTargetDefaults.inputs) {
      for (const webpackExecutorTargetDefaultsInput of webpackExecutorTargetDefaults.inputs) {
        if (typeof webpackExecutorTargetDefaultsInput === 'object') {
          const [key, value] = Object.entries(
            webpackExecutorTargetDefaultsInput
          )[0];

          if (key === 'env' && value === mfEnvVar) {
            mfEnvVarExists = true;
            break;
          }
        }
      }
    } else {
      nxJson.targetDefaults[webpackExecutor].inputs = [];
    }
  } else if (!nxJson.targetDefaults[webpackExecutor]) {
    nxJson.targetDefaults[webpackExecutor] = {
      inputs: [],
    };
  } else if (!nxJson.targetDefaults) {
    nxJson.targetDefaults = {
      [webpackExecutor]: {
        inputs: [],
      },
    };
  }

  if (!mfEnvVarExists) {
    nxJson.targetDefaults[webpackExecutor].inputs.push({ env: mfEnvVar });
  }

  updateNxJson(tree, nxJson);
}
