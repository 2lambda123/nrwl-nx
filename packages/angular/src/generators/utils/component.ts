import {
  joinPathFragments,
  names,
  normalizePath,
  readProjectConfiguration,
  readWorkspaceConfiguration,
  Tree,
} from '@nrwl/devkit';

export type ComponentGenerationOptions = {
  name: string;
  flat?: boolean;
  path?: string;
  project?: string;
  type?: string;
};
export type ComponentFileInfo = {
  componentDirectory: string;
  componentFileName: string;
  componentFilePath: string;
};

export function getComponentFileInfo(
  tree: Tree,
  options: ComponentGenerationOptions
): ComponentFileInfo {
  const project =
    options.project ?? readWorkspaceConfiguration(tree).defaultProject;
  const { root, sourceRoot, projectType } = readProjectConfiguration(
    tree,
    project
  );
  const { fileName: normalizedName } = names(options.name);

  const componentFileName = `${normalizedName}.${
    options.type ? names(options.type).fileName : 'component'
  }`;

  const projectSourceRoot = sourceRoot ?? joinPathFragments(root, 'src');
  const path =
    options.path ??
    joinPathFragments(
      projectSourceRoot,
      projectType === 'application' ? 'app' : 'lib'
    );
  const componentDirectory = options.flat
    ? normalizePath(path)
    : joinPathFragments(path, normalizedName);

  const componentFilePath = joinPathFragments(
    componentDirectory,
    `${componentFileName}.ts`
  );

  return { componentDirectory, componentFileName, componentFilePath };
}
