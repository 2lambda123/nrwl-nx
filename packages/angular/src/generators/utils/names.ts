import { normalizePath } from '@nrwl/devkit';

export type NameInfo = { name: string; path: string };

export function parseName(rawName: string): NameInfo {
  const parsedName = normalizePath(rawName).split('/');
  const name = parsedName.pop();
  const path = parsedName.join('/');

  return { name, path };
}
