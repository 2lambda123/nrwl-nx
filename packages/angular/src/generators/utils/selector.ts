import type { Tree } from '@nrwl/devkit';
import { names, readNxJson } from '@nrwl/devkit';

export function buildSelector(
  tree: Tree,
  name: string,
  prefix: string | undefined,
  projectPrefix: string | undefined,
  casing: keyof Pick<ReturnType<typeof names>, 'fileName' | 'propertyName'>
): string {
  let selector = name;
  prefix ??= projectPrefix ?? readNxJson(tree)?.npmScope;
  if (prefix) {
    selector = `${prefix}-${selector}`;
  }

  return names(selector)[casing];
}
