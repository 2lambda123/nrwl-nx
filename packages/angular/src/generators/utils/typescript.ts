import { Tree } from '@nrwl/devkit';

export function getRootTsConfigPath(tree: Tree): string {
  for (const path of ['tsconfig.base.json', 'tsconfig.json']) {
    if (tree.exists(path)) {
      return path;
    }
  }

  throw new Error(
    'Could not find root tsconfig file. Tried with "tsconfig.base.json" and "tsconfig.json".'
  );
}
