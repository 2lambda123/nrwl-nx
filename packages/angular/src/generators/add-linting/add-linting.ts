import {
  GeneratorCallback,
  joinPathFragments,
  Tree,
  updateJson,
} from '@nrwl/devkit';
import { Linter, lintProjectGenerator } from '@nrwl/linter';
import { mapLintPattern } from '@nrwl/linter/src/generators/lint-project/lint-project';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import { join } from 'path';
import { getGeneratorDirectoryForInstalledAngularVersion } from '../utils/version-utils';
import { addAngularEsLintDependencies } from './lib/add-angular-eslint-dependencies';
import { extendAngularEslintJson } from './lib/create-eslint-configuration';
import type { AddLintingGeneratorSchema } from './schema';

export async function addLintingGenerator(
  tree: Tree,
  options: AddLintingGeneratorSchema
): Promise<GeneratorCallback> {
  const generatorDirectory =
    getGeneratorDirectoryForInstalledAngularVersion(tree);
  if (generatorDirectory) {
    let previousGenerator = await import(
      join(__dirname, generatorDirectory, 'add-linting')
    );
    await previousGenerator.default(tree, options);
    return;
  }

  const tasks: GeneratorCallback[] = [];
  const rootProject = options.projectRoot === '.' || options.projectRoot === '';
  const lintTask = await lintProjectGenerator(tree, {
    linter: Linter.EsLint,
    project: options.projectName,
    tsConfigPaths: [
      joinPathFragments(options.projectRoot, 'tsconfig.app.json'),
    ],
    unitTestRunner: options.unitTestRunner,
    eslintFilePatterns: [
      mapLintPattern(options.projectRoot, 'ts', rootProject),
      mapLintPattern(options.projectRoot, 'html', rootProject),
    ],
    setParserOptionsProject: options.setParserOptionsProject,
    skipFormat: true,
    rootProject: rootProject,
  });
  tasks.push(lintTask);

  updateJson(
    tree,
    joinPathFragments(options.projectRoot, '.eslintrc.json'),
    (json) => extendAngularEslintJson(json, options)
  );

  if (!options.skipPackageJson) {
    const installTask = await addAngularEsLintDependencies(tree);
    tasks.push(installTask);
  }

  return runTasksInSerial(...tasks);
}

export default addLintingGenerator;
