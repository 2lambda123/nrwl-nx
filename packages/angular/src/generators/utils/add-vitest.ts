import {
  addDependenciesToPackageJson,
  ensurePackage,
  joinPathFragments,
  readJson,
  type Tree,
  updateJson,
} from '@nx/devkit';
import { analogVitestAngular, nxVersion } from '../../utils/versions';

export type AddVitestOptions = {
  name: string;
  projectRoot: string;
  skipPackageJson: boolean;
  strict: boolean;
};

export async function addVitest(
  tree: Tree,
  options: AddVitestOptions
): Promise<void> {
  if (!options.skipPackageJson) {
    process.env.npm_config_legacy_peer_deps ??= 'true';

    addDependenciesToPackageJson(
      tree,
      {},
      {
        '@analogjs/vitest-angular': analogVitestAngular,
        '@analogjs/vite-plugin-angular': analogVitestAngular,
      },
      undefined,
      true
    );
  }

  const { createOrEditViteConfig, viteConfigurationGenerator } = ensurePackage<
    typeof import('@nx/vite')
  >('@nx/vite', nxVersion);

  const relativeTestSetupPath = joinPathFragments('src', 'test-setup.ts');
  const isEsmImplicit = readJson(tree, 'package.json').type === 'module';

  createOrEditViteConfig(
    tree,
    {
      project: options.name,
      includeLib: false,
      includeVitest: true,
      inSourceTests: false,
      imports: [`import angular from '@analogjs/vite-plugin-angular'`],
      plugins: ['angular()'],
      setupFile: relativeTestSetupPath,
      useEsmExtension: !isEsmImplicit,
    },
    true
  );

  const setupFile = joinPathFragments(
    options.projectRoot,
    relativeTestSetupPath
  );
  if (!tree.exists(setupFile)) {
    tree.write(
      setupFile,
      `import '@analogjs/vitest-angular/setup-zone';

import {
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting,
} from '@angular/platform-browser-dynamic/testing';
import { getTestBed } from '@angular/core/testing';

getTestBed().initTestEnvironment(
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting()
);
`
    );
  }

  await viteConfigurationGenerator(tree, {
    project: options.name,
    newProject: true,
    uiFramework: 'angular',
    includeVitest: true,
    testEnvironment: 'jsdom',
  });

  updateJson(
    tree,
    joinPathFragments(options.projectRoot, 'tsconfig.spec.json'),
    (tsConfig) => {
      tsConfig.files = [...(tsConfig.files ?? []), relativeTestSetupPath];
      return tsConfig;
    }
  );

  updateJson(
    tree,
    joinPathFragments(options.projectRoot, 'tsconfig.app.json'),
    (tsConfig) => {
      tsConfig.exclude = [...(tsConfig.exclude ?? []), relativeTestSetupPath];
      return tsConfig;
    }
  );
}
