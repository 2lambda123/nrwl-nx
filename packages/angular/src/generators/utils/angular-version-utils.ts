import type { Tree } from '@nrwl/devkit';
import { readJson } from '@nrwl/devkit';
import { clean, coerce, major } from 'semver';
import { angularVersion } from '../../utils/versions';

export function getGeneratorDirectoryForInstalledAngularVersion(
  tree: Tree
): string | null {
  const majorAngularVersion = getInstalledAngularMajorVersion(tree);

  const directoryDictionary = {
    14: 'angular-v14',
  };

  return directoryDictionary[majorAngularVersion] ?? null;
}

export function getInstalledAngularVersion(tree: Tree): string {
  const pkgJson = readJson(tree, 'package.json');
  const installedAngularVersion =
    pkgJson.dependencies && pkgJson.dependencies['@angular/core'];

  if (
    !installedAngularVersion ||
    installedAngularVersion === 'latest' ||
    installedAngularVersion === 'next'
  ) {
    return clean(angularVersion) ?? coerce(angularVersion).version;
  }

  return (
    clean(installedAngularVersion) ?? coerce(installedAngularVersion).version
  );
}

export function getInstalledAngularMajorVersion(tree: Tree): number {
  return major(getInstalledAngularVersion(tree));
}

export function getInstalledAngularVersionInfo(tree: Tree) {
  const installedVersion = getInstalledAngularVersion(tree);

  return {
    version: installedVersion,
    major: major(installedVersion),
  };
}

export function getInstalledPackageVersionInfo(tree: Tree, pkgName: string) {
  try {
    const version =
      readJson(tree, 'package.json').dependencies?.[pkgName] ??
      readJson(tree, 'package.json').devDependencies?.[pkgName];

    return { major: major(coerce(version)), version };
  } catch {
    return null;
  }
}
