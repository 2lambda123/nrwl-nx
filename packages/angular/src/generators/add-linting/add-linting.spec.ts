import 'nx/src/internal-testing-utils/mock-project-graph';
import {
  ProjectConfiguration,
  Tree,
  addProjectConfiguration,
  readJson,
  updateJson,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import * as linter from '@nx/eslint';
import { addLintingGenerator } from './add-linting';
import { eslint9__eslintVersion } from '@nx/eslint/src/utils/versions';

jest.mock('nx/src/devkit-internals', () => ({
  ...jest.requireActual<any>('nx/src/devkit-internals'),
  readModulePackageJson: jest.fn().mockImplementation((module, extras) => {
    if (module === 'eslint') {
      return {
        packageJson: {
          version: '9.8.0',
        },
      };
    } else {
      return jest
        .requireActual<any>('nx/src/devkit-internals')
        .readModulePackageJson(module, extras);
    }
  }),
}));

describe('addLinting generator', () => {
  let tree: Tree;
  const appProjectName = 'ng-app1';
  const appProjectRoot = `apps/${appProjectName}`;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    addProjectConfiguration(tree, appProjectName, {
      root: appProjectRoot,
      prefix: 'myOrg',
      projectType: 'application',
      targets: {},
    } as ProjectConfiguration);
  });

  it('should invoke the lintProjectGenerator', async () => {
    jest.spyOn(linter, 'lintProjectGenerator');

    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
    });

    expect(linter.lintProjectGenerator).toHaveBeenCalled();
  });

  it('should add the Angular specific EsLint devDependencies', async () => {
    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
    });

    const { devDependencies } = readJson(tree, 'package.json');
    expect(devDependencies['@angular-eslint/eslint-plugin']).toBeDefined();
    expect(
      devDependencies['@angular-eslint/eslint-plugin-template']
    ).toBeDefined();
    expect(devDependencies['@angular-eslint/template-parser']).toBeDefined();
  });

  it('should use flat config and install correct dependencies when using it', async () => {
    updateJson(tree, 'package.json', (json) => {
      json.devDependencies['eslint'] = eslint9__eslintVersion;
      return json;
    });

    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
    });

    const { devDependencies } = readJson(tree, 'package.json');
    expect(devDependencies['@typescript-eslint/utils']).toMatchInlineSnapshot(
      `"^8.0.0"`
    );
  });

  it('should correctly generate the .eslintrc.json file', async () => {
    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
    });

    const eslintConfig = readJson(tree, `${appProjectRoot}/.eslintrc.json`);
    expect(eslintConfig).toMatchSnapshot();
  });

  it('should not touch the package.json when run with `--skipPackageJson`', async () => {
    let initialPackageJson;
    updateJson(tree, 'package.json', (json) => {
      json.dependencies = {};
      json.devDependencies = {};
      initialPackageJson = json;

      return json;
    });

    await addLintingGenerator(tree, {
      prefix: 'myOrg',
      projectName: appProjectName,
      projectRoot: appProjectRoot,
      skipFormat: true,
      skipPackageJson: true,
    });

    const packageJson = readJson(tree, 'package.json');
    expect(packageJson).toEqual(initialPackageJson);
  });
});
