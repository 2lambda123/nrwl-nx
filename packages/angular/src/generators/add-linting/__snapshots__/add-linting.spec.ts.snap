// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addLinting generator should correctly generate the .eslintrc.json file 1`] = `
Object {
  "ignorePatterns": Array [
    "!**/*",
  ],
  "overrides": Array [
    Object {
      "extends": Array [
        "plugin:@nrwl/nx/angular",
        "plugin:@angular-eslint/template/process-inline-templates",
        "../../.eslintrc.json",
      ],
      "files": Array [
        "*.ts",
      ],
      "rules": Object {
        "@angular-eslint/component-selector": Array [
          "error",
          Object {
            "prefix": "my-org",
            "style": "kebab-case",
            "type": "element",
          },
        ],
        "@angular-eslint/directive-selector": Array [
          "error",
          Object {
            "prefix": "myOrg",
            "style": "camelCase",
            "type": "attribute",
          },
        ],
      },
    },
    Object {
      "extends": Array [
        "plugin:@nrwl/nx/angular-template",
        "../../.eslintrc.json",
      ],
      "files": Array [
        "*.html",
      ],
      "rules": Object {},
    },
  ],
}
`;
